### PR TITLE
chore(vdev): Rework support for integration tests

### PR DESCRIPTION
**Pull request by bruceg**
This reworks the support in `vdev` for integration testing, starting from the integration test environments (#15232 through #15259) and refactored the program code into vdev, and incrementally moving all of the test framework inside of the tool, leaving only config files outside of the `vdev` directory, and then refactoring the result.

As of this PR, several of the integration tests run properly, but not all, and they are not used in the CI process. I will submit a second PR following this one to clean up the actual tests, as that should not impact the test runner. I am putting this out now so the runner portion can get review while I work through the individual tests.

Notably, the following tests work out of the box:
* `amqp`
* `apex`
* `azure`
* `chronicle`
* `clickhouse`
* `elasticsearch`
* `http-client`
* `kafka`
* `nats`
* `redis`

You can run them with:
```shell
cargo vdev integration test http-client
```
or, as individual steps to enable faster repeat testing:
```shell
cargo vdev integration start http-client latest
cargo vdev integration test http-client latest
cargo vdev integration stop http-client
```

### CODE REVIEW

**Comment by spencergilbert** [comment]
Is there any documentation around what goes in `compose` vs `test` for the config files?

**Comment by bruceg** [comment]
> Is there any documentation around what goes in `compose` vs `test` for the config files?

Ah, no. The `compose.yaml` file is a `docker-compose` file, which is [apparently the new standard name](https://github.com/compose-spec/compose-spec/blob/master/spec.md#compose-file). The other `test.yaml` file configures how the test is run outside of compose.

File: scripts/integration/datadog-traces/compose.yaml
Comment (spencergilbert) [inline scripts/integration/datadog-traces/compose.yaml:1]: I find it interesting that this is separate from the `agent` test above, when it should be the same component?

@neuronull are there reasons we can't reduce these to a single test setup?

File: scripts/integration/prometheus/compose.yaml
Comment (spencergilbert) [inline scripts/integration/prometheus/compose.yaml:18]: Noting that I'm real annoyed that we involve influx in our prometheus tests...

File: vdev/src/commands/integration/mod.rs
Comment (spencergilbert) [inline vdev/src/commands/integration/mod.rs:5]: We had briefly discussed moving these out of `scripts/`, is that still desired/planned later or is it not feasible?

**Review by spencergilbert — COMMENTED** [review]
Generally happy with the changes, I'll make some time to play with it locally as well

File: scripts/integration/README.md
Comment (bruceg) [reply scripts/integration/README.md:1]: Correct. I had created the file just for a convenience when running for loops over the tests.

File: scripts/integration/aws/compose.yaml
Comment (bruceg) [reply scripts/integration/aws/compose.yaml:23]: I… don't know? It is removed from the compose file compared to the original `docker-compose.aws.yaml` file. I'll drop it.

File: scripts/integration/axiom/compose.yaml
Comment (bruceg) [reply scripts/integration/axiom/compose.yaml:5]: This is a bit of a wart that I will be resolving with a follow-up PR. The variable `AXIOM_VERSION` is produced from the `version` value in the `matrix` in `test.yaml`, where `AXIOM` is created from the test name. This isn't actually the "axiom" version, it's just a version field on the axiom test. Terribly confusing. This will be changing so I can name it as the postgres version specifically.

File: scripts/integration/chronicle/compose.yaml
Comment (bruceg) [reply scripts/integration/chronicle/compose.yaml:9]: Indeed, the original files had _four_ `..`, which led to the wrong path. The original files have `${PWD}` but the new setup executes with the current working directory set to the individual integration, with the idea that you could reference files within that directory without a path.

Hmm, having just written that, I see that that particular one just needs to be `../chroniclepub.pem`, which is much less easy to get wrong. The rest reference files in `tests/data`. I intend to move these configs out of `scripts` in a follow-up PR, so I'll look into cleaning up the relative paths at that point.

File: vdev/src/commands/integration/mod.rs
Comment (bruceg) [reply vdev/src/commands/integration/mod.rs:5]: My plan is to get all of the integration tests working through vdev in a follow-up PR (work in progress), and then to refactor them and all the configuration files into a more suitable location.

File: scripts/integration/prometheus/compose.yaml
Comment (spencergilbert) [reply scripts/integration/prometheus/compose.yaml:18]: Ah, if it's just for remote write - it is supported, I think it's just behind a config flag or something similar 👍

File: scripts/integration/README.md
Comment (neuronull) [inline scripts/integration/README.md:24]: I haven't inspected the code deeply yet, but I pulled the PR and just ran the `show` command.
Maybe I'm being dense but it isn't super clear to me what the "environment" is? Is it just a container version?

Also, I had to read this twice to discern that the `cargo` part of the command strings together the start, test, and stop subcommands of vdev for you. Might be worth calling out?

**Review by neuronull — COMMENTED** [review]
tested it out locally 👍

File: scripts/integration/README.md
Comment (bruceg) [reply scripts/integration/README.md:24]: The "environment" is a little… flexible. It is a product of all the fields listed in the `matrix` of the `test.yaml` file. For most of them (all currently?), it is the version or tag of the container, but that is not necessarily its only use.

File: vdev/src/commands/integration/show.rs
Comment (bruceg) [reply vdev/src/commands/integration/show.rs:8]: Agreed, will fix.

File: scripts/integration/README.md
Comment (spencergilbert) [reply scripts/integration/README.md:24]: Perhaps there a better term than environment we could think up?

**Comment by neuronull** [comment]
Perhaps a dumb question... should this PR be running integration tests in CI ?? Or are we not "switching over" yet.

File: scripts/integration/README.md
Comment (spencergilbert) [reply scripts/integration/README.md:1]: 👍 leaving this open for other reviewers

File: scripts/integration/prometheus/compose.yaml
Comment (spencergilbert) [reply scripts/integration/prometheus/compose.yaml:18]: Regardless, unrelated to the changes here

**Review by spencergilbert — APPROVED** [review]

File: scripts/integration/README.md
Comment (neuronull) [reply scripts/integration/README.md]: I suppose keeping the name "environment" is ok, but maybe we could add a note in the README wrt to the clarification Bruce just provided. Since just looking at the output of the show command, and seeing a jumble of numbers and "latest" , but all currently having a single "environment" could be puzzling. It was for me at least.

```
Integration Name  Environment Name(s)
----------------  -------------------
amqp              3.8
apex              latest
aws               latest
axiom             13-alpine
azure             3.14.0
chronicle         latest
clickhouse        19
datadog-agent     7 (active)
datadog-traces    latest
elasticsearch     7.13.1
eventstoredb      latest
gcp               latest
http-client       latest
humio             1.13.1
influxdb          1.8
kafka             latest
logstash          7.12.1
loki              2.4.1
mongodb           4.2.10
nats              latest
nginx             1.19.4
opentelemetry     0.56.0
postgres          13.1
prometheus        1.8
pulsar            latest
redis             6-alpine
shutdown          latest
splunk            8.2.4
```

File: scripts/integration/datadog-traces/compose.yaml
Comment (neuronull) [reply scripts/integration/datadog-traces/compose.yaml:1]: I'm not really clear on what the `version` field of the yaml file does. hmmm

But to the question of why there is a separate integration suite for `datadog-traces`, I think the main reason is that the agent service has different configurations for the datadog traces test. 

Also, we would need to have separate args to nextest passed for the source vs sink .

File: scripts/integration/README.md
Comment (neuronull) [reply scripts/integration/README.md]: Seems that Bruce is saying that isn't always necessarily going to be the case though.

File: scripts/integration/datadog-traces/compose.yaml
Comment (spencergilbert) [reply scripts/integration/datadog-traces/compose.yaml:1]: Doh - sorry yeah, it wasn't specific to that version line 😂 

It seemed like the other test was doing all "sources and sinks" that are Datadog (except for this trace one), I may have misread it.

Comment (neuronull) [reply scripts/integration/datadog-traces/compose.yaml:1]: also a side note partially for myself- on my TODO list still is to update the scope of the integration test files to only run the integration tests.
I'll wait until after this is merged in for that.

Comment (bruceg) [reply scripts/integration/datadog-traces/compose.yaml:1]: Good plan, however most of these integration test configs do already run only a limited set of tests.

Comment (bruceg) [reply scripts/integration/datadog-traces/compose.yaml:1]: > I'm not really clear on what the `version` field of the yaml file does. hmmm

The `version` field is the version of the compose standard to use.

**Pull request merged**
